pkg/resource/provider/main.go

```diff
@@ -52,12 +55,27 @@ func Main(name string, provMaker func(*HostClient) (pulumirpc.ResourceProviderSe
 		if err != nil {
 			return fmt.Errorf("fatal: could not connect to host RPC: %v", err)
 		}
+
+		// If we have a host cancel our cancellation context if it fails the healthcheck
+		if host != nil {
+			ctx, cancel := context.WithCancel(context.Background())
+			// map the context Done channel to the rpcutil boolean cancel channel
+			cancelChannel = make(chan bool)
+			go func() {
+				<-ctx.Done()
+				close(cancelChannel)
+			}()
+			err = rpcutil.Healthcheck(ctx, args[0], 5*time.Minute, cancel)
+			if err != nil {
+				return fmt.Errorf("could not start health check host RPC server: %w", err)
+			}
+		}
```

Comment:
style nit: it doesn't follow for me here is that we check host != nil then don't use host in the consequent. If host == nil is a possibility (even with err == nil), is that an error we should handle separately?

Might be cleaner to do: 

```go
if host == nil { 
  /* error */ 
}
// host != nil case, unindented
```

Reply:
It's not an error for host to be nil. That's what we get if we're doing an provider attach.

Reply:
Although the host == nil check shouldn't actually be needed, we assume NewHostClient won't return nil if err was nil